src/qe/mbp/mbp_term_graph.cpp

```diff
@@ -348,29 +598,68 @@ namespace mbp {
         merge(*internalize_term(a1), *internalize_term(a2));
         merge_flush();
         SASSERT(m_merge.empty());
+        if (!m_explicit_eq) return;
+        expr_ref eq(m.mk_eq(a1, a2), m);
+        term* res = get_term(eq);
+        if (!res)
+          mk_term(eq);
     }
 
-    void term_graph::internalize_lit(expr* lit) {
-        expr *e1 = nullptr, *e2 = nullptr, *v = nullptr;
-        if (m.is_eq (lit, e1, e2)) {
-            internalize_eq (e1, e2);
-        }
-        else {
-            internalize_term(lit);
+    void term_graph::internalize_distinct(expr *d) {
+      app *a = to_app(d);
+      ptr_vector<term> ts(a->get_decl()->get_arity());
+      auto tsit = ts.begin();
+      for (auto arg : *a) {
+        *tsit = internalize_term(arg);
+        tsit++;
+      }
+      m_add_deq(ts);
+      m_deq_distinct.push_back(ts);
+      if (!m_explicit_eq) return;
+      term* t  = get_term(d);
+      if (!t) mk_term(d);
+    }
+
+    void term_graph::internalize_deq(expr *a1, expr *a2) {
+      // TODO: what do not add disequalities of interpreted terms? (e.g. 1 != 2)
```

Comment:
Was this comment mine? I don't think these disequalities are added, adding them could be implemented as a plugin that finds if a class equals some value.